### PR TITLE
refactor `fileio.c`

### PR DESCRIPTION
**Pull request by lewis6991**
## Problem

`buf_write()` is:
* too long (1478 LOC)
* contains many `goto` statements
* contains too many `ifdefs`

## Solution

* Break up `buf_write()` (\* currently on 883 lines).
* Tidy up the `ifdefs`

### CODE REVIEW

File: src/nvim/fileio.c [hunk 1]
```diff
@@ -2079,6 +2094,658 @@ char *new_file_message(void)
...
+  } else {
+    apply_autocmds_exarg(EVENT_FILEWRITEPOST, fname, fname,
+                         false, curbuf, eap);
+  }
+
+  // restore curwin/curbuf and a few other things
+  aucmd_restbuf(&aco);
+}
+
+static void set_err_num(Error_T *e, const char *num, const char *msg)
```
Comment (bfredl) [inline src/nvim/fileio.c:2295]: These can be written as plain assignments i e `err = err_num(num, msg)` at the call site by returning the struct, which makes it immediately clear we are overwriting the entire struct (even if a subset of fields are specificed)

Comment (lewis6991) [reply src/nvim/fileio.c]: Done

File: src/nvim/os/fs.h [hunk 2]
```diff
@@ -2,8 +2,21 @@
...
+# define OS_GET_ACL(aclent, fname) { (aclent) = os_get_acl(fname); }
+# define OS_FREE_ACL(aclent) os_free_acl(aclent)
+# define OS_SET_ACL(fname, aclent) os_set_acl((fname), (aclent))
+#else
+# define OS_NEW_ACL(aclent)
+# define OS_GET_ACL(aclent, fname)
+# define OS_FREE_ACL(aclent)
+# define OS_SET_ACL(fname, aclent)
+#endif
+
```
Comment (zeertzjq) [inline src/nvim/os/fs.h:19]: `os_get_acl()` and `os_set_acl()` are actually always no-op, so this just adds another layer of redundant wrapper.

Comment (zeertzjq) [reply src/nvim/os/fs.h:19]: Maybe it's better to make these functions always available?

Comment (lewis6991) [reply src/nvim/os/fs.h:19]: If they are always no-op, then why do they exist?

Comment (zeertzjq) [reply src/nvim/os/fs.h:19]: For future implementation, I guess.

Comment (lewis6991) [reply src/nvim/os/fs.h]: ok fair enough. Let's just expose them.

**Pull request merged**